src/opera/csar.py

```diff
@@ -1,13 +1,207 @@
-    with open("{}.deploy".format(name)) as fd:
-        return json.load(fd)["name"]
+class ToscaCsar(ABC):
+    """
+    The TOSCA Cloud Service Archive.
+
+    Does not support the service metadata format.
+    """
+    METADATA_DIRNAME = "TOSCA-Metadata"
+    METADATA_FILENAME = "TOSCA.meta"
+
+    @classmethod
+    def load(cls, path_string: str) -> "ToscaCsar":
+        logger.info("Loading CSAR from %s", path_string)
+        path = pathlib.Path(path_string)
+        if path.is_file():
+            try:
+                logger.debug("Attempting to use a zipped CSAR.")
+                csar: ToscaCsar = ZipToscaCsar(path_string)
+            except zipfile.BadZipFile:
+                logger.debug("File not a zip file, assuming a bare service template.")
+                csar = VirtualToscaCsar(path_string)
+        elif path.is_dir():
+            logger.debug("Using a non-standard directory CSAR.")
+            csar = DirectoryToscaCsar(path_string)
+        else:
+            raise CsarValidationError("The CSAR is neither a file nor a directory.", "6.1 (xopera extended)")
+
+        csar.validate()
+        return csar
+
+    @abstractmethod
+    def members(self) -> List[pathlib.PurePath]:
+        """
+        Get a list of files and directories with paths relative to the root of the CSAR. Absolute paths not allowed.
+        """
+
+    @abstractmethod
+    def open_member(self, path: pathlib.PurePath) -> IO:
+        """
+        Opens a read-only stream for the file at the specified path.
+        """
+
+    @abstractmethod
+    def _member_is_dir(self, member: pathlib.PurePath) -> bool:
+        """
+        Assumes the path, relative to the CSAR root, exists, and returns whether it is a directory.
+        Symlinks are resolved.
+        """
+
+    @abstractmethod
+    def _member_is_file(self, member: pathlib.PurePath) -> bool:
+        """
+        Assumes the path, relative to the CSAR root, exists, and returns whether it is a regular file.
+        Symlinks are resolved.
+        """
+
+    def get_service_template(self) -> pathlib.PurePath:
+        """
+        Returns the service template file path. Assumes it exists.
+        """
+        relative_root = pathlib.PurePath(".")
+        return next(f for f in self.members() if f.parent == relative_root and self._member_is_file(f))
+
+    def validate(self):
+        """
+        Validates the CSAR structure and throws an exception if it does not conform to the specification.
+        Does not check service templates or dependencies.
+        """
+        logger.debug("Validating CSAR.")
+        files = self.members()
+
+        relative_root = pathlib.PurePath(".")
+        if any(f for f in files if
+               f.parent == relative_root
+               and f.name == ToscaCsar.METADATA_DIRNAME
+               and self._member_is_dir(f)):
+            raise UnsupportedToscaFeatureError("xopera does not support CSAR structures with a metadata directory.")
```

Comment:
`xOpera`

Reply:
This applies to all text occurences in comments, logger outputs and error messages